BranchSDK/src/BranchIO/Event/BaseEvent.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2019 Branch Metrics, Inc.
+
+#ifndef BRANCHIO_EVENT_BASEEVENT_H_
+#define BRANCHIO_EVENT_BASEEVENT_H_
+
+#include <string>
+#include "BranchIO/Defines.h"
+#include "BranchIO/fwd.h"
+#include "BranchIO/PropertyManager.h"
+
+namespace BranchIO {
+
+typedef std::string CurrencyType;
+
+/**
+ * Base class for Branch events for tracking and analytical purpose.
+ */
+class BRANCHIO_DLL_EXPORT BaseEvent : public PropertyManager {
+ public:
+    using PropertyManager::toString;
+
+    virtual ~BaseEvent();
+
+    /**
+     * @return the Event Name.
+     */
+    virtual const std::string &name() const;
+
+    /**
+     * Adds a custom data property associated with this Branch Event.
+     * @param propertyName Name of the custom property
+     * @param propertyValue Value of the custom property
+     * @return This object for chaining builder methods
+     */
+    virtual BaseEvent& addCustomDataProperty(const std::string &propertyName, const std::string &propertyValue);
+
+    /**
+     * Get the custom data properties associated with this Branch Event.
+     * @return the Custom Data for the event.
+     */
+    virtual const JSONObject::Ptr getCustomData() const;
+
+ public:
+    /**
+     * @return the API Endpoint.
+     */
+    virtual Defines::APIEndpoint getAPIEndpoint() const;
+
+    /**
+     * Prepare the event package for transmission.
+     * @param packagingInfo Context for packaging events
+     * @param jsonPackage result of the packaging process
+     * @todo(andyp) Revisit Scope
```

Comment:
👍 re revisiting scope. I'd argue `packageV1Event` and `packageV2Event` should be virtual, not `package `

Reply:
Talked in person, it makes more sense to override `package` if, for example, we are implementing some funky request that's neither v1 or v2, so we are keeping implementation as is.